### ISSUE TITLE
Python and pip paths are not set to virtualenv

### ISSUE THREAD

**Issue by tecoholic** [jdx/mise#3069]
**Describe the bug**

Python Virtualenv is not activated.

**To Reproduce**

```sh
~/code ❯ mkdir test
~/code ❯ cd test/
~/code/test ❯ mise use python@3.11
mise ~/code/test/mise.toml tools: python@3.11.9

~/code/test ❯ nvim mise.toml
mise ERROR Config file ~/code/test/mise.toml is not trusted.
Trust it with `mise trust`.
mise ERROR Run with --verbose or MISE_VERBOSE=1 for more information

~/code/test took 16s ❯ mise trust
mise trusted /home/tecoholic/code/test/mise.toml
mise creating venv at: ~/code/test/.venv

~/code/test via  v3.11.9 (.venv) ❯ cat mise.toml
[tools]
python = "3.11"

[env]
_.python.venv = { path = ".venv", create = true }

~/code/test via  v3.11.9 (.venv) took 2s ❯ which python
/home/tecoholic/.local/share/mise/shims/python

~/code/test via  v3.11.9 (.venv) ❯ python --version
Python 3.11.9
```

Checking the interpreter path using `which`. 
```sh
~/code/test via  v3.11.9 (.venv) ❯ which python
/home/tecoholic/.local/share/mise/shims/python

~/code/test via  v3.11.9 (.venv) ❯ which pip
/home/tecoholic/.local/share/mise/shims/pip
```

After activating the virtualenv
```sh
~/code/test via  v3.11.9 (.venv) ❯ source .venv/bin/activate

~/code/test via  v3.11.9 (.venv) ❯ which python
/home/tecoholic/code/test/.venv/bin/python

~/code/test via  v3.11.9 (.venv) ❯ which pip
/home/tecoholic/code/test/.venv/bin/pip
```

However, the [docs tell me shims break which](https://mise.jdx.dev/dev-tools/shims.html#which). So I am not sure if I can fully trust what I am seeing above. Maybe the shims point to the right interpreter, after all?

So, I used `mise which`. It returns the install path instead of the virtualenv path, both before and after running `source .venv/bin/activate`. So it is even more confusing as to which interpreter is going to used when I run python. 
```sh
~/code/test via  v3.11.9 (.venv) ❯ mise which python
/home/tecoholic/.local/share/mise/installs/python/3.11.9/bin/python

~/code/test via  v3.11.9 (.venv) ❯ mise which pip
/home/tecoholic/.local/share/mise/installs/python/3.11.9/bin/pip
```

Finally, I confirmed it with `sys.executable`

```sh
# After simply navigating into the folder
~/code/test via  v3.11.9 (.venv) ❯ python -c "import sys; print(sys.executable)"
/home/tecoholic/.local/share/mise/installs/python/3.11.9/bin/python

# Manually activating and checking
~/code/test via  v3.11.9 (.venv) ❯ source .venv/bin/activate

~/code/test via  v3.11.9 (.venv) ❯ python -c "import sys; print(sys.executable)"
/home/tecoholic/code/test/.venv/bin/python
```

**Expected behavior**

After the virtualenv is created, the python interpretor and pip should be the one in the virtual environment. Currently, it requires manually activating the environment for them to be properly set, despite the prompt saying otherwise.

**`mise doctor` output**

```text
version: 2024.11.15 linux-x64 (d3d8dc0 2024-11-16)
activated: yes
shims_on_path: yes

build_info:
  Target: x86_64-unknown-linux-gnu
  Features: DEFAULT, NATIVE_TLS, OPENSSL
  Built: Sat, 16 Nov 2024 16:23:17 +0000
  Rust Version: rustc 1.82.0 (f6e511eec 2024-10-15)
  Profile: release

shell:
  /bin/bash
  GNU bash, version 5.2.32(1)-release (x86_64-pc-linux-gnu)
  Copyright (C) 2022 Free Software Foundation, Inc.
  License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>

  This is free software; you are free to change and redistribute it.
  There is NO WARRANTY, to the extent permitted by law.

dirs:
  data: ~/.local/share/mise
  config: ~/.config/mise
  cache: ~/.cache/mise
  state: ~/.local/state/mise
  shims: ~/.local/share/mise/shims

config_files:
  ~/.config/mise/config.toml
  ~/code/test/mise.toml

backends:
  aqua
  asdf
  cargo
  core
  go
  npm
  pipx
  spm
  ubi
  vfox
  unknown

plugins:
  tutor  git@github.com:tecoholic/asdf-tutor.git#9e92340
  yarn   https://github.com/mise-plugins/asdf-yarn.git#74ea3b9

toolset:
  core:python@3.11.9
  core:node@22.7.0

env_vars:
  MISE_SHELL=bash

settings:
  activate_aggressive = false
  all_compile = false
  always_keep_download = false
  always_keep_install = false
  asdf_compat = false
  cache_prune_age = "30d"
  ci = false
  color = true
  debug = false
  disable_backends = []
  disable_default_registry = false
  disable_hints = []
  disable_tools = []
  experimental = true
  fetch_remote_versions_cache = "1h"
  fetch_remote_versions_timeout = "10s"
  go_default_packages_file = "~/.default-go-packages"
  go_download_mirror = "https://dl.google.com/go"
  go_repo = "https://github.com/golang/go"
  go_set_gopath = false
  go_set_goroot = true
  go_skip_checksum = false
  http_timeout = "30s"
  jobs = 4
  legacy_version_file = true
  legacy_version_file_disable_tools = []
  libgit2 = true
  lockfile = false
  log_level = "info"
  not_found_auto_install = true
  paranoid = false
  pin = false
  plugin_autoupdate_last_check_duration = "7d"
  python_compile = true
  quiet = false
  raw = false
  task_timings = false
  trace = false
  trusted_config_paths = []
  unix_default_file_shell_args = ["sh"]
  unix_default_inline_shell_args = [
      "sh",
      "-c",
  ]
  use_file_shell_for_executable_tasks = false
  use_versions_host = true
  verbose = false
  windows_default_file_shell_args = [
      "cmd",
      "/c",
  ]
  windows_default_inline_shell_args = [
      "cmd",
      "/c",
  ]
  windows_executable_extensions = [
      "exe",
      "bat",
      "cmd",
      "com",
      "ps1",
      "vbs",
  ]
  yes = false

  [cargo]
  binstall = true

  [node]

  [npm]
  bun = false

  [pipx]
  uvx = false

  [python]
  compile = true
  default_packages_file = "~/.default-python-packages"
  pyenv_repo = "https://github.com/pyenv/pyenv.git"
  venv_auto_create = false
  venv_stdlib = false

  [ruby]
  default_packages_file = "~/.default-gems"
  ruby_build_repo = "https://github.com/rbenv/ruby-build.git"
  ruby_install = false
  ruby_install_repo = "https://github.com/postmodern/ruby-install.git"

  [status]
  missing_tools = "if_other_versions_installed"
  show_env = false
  show_tools = false
No warnings found
No problems found

```

**Additional context**

This feels like a regression in the last couple of releases, because, I don't remember having to manually activate the virtual env before.

**Comment by syhol** [comment]
Seems like you have both modes activated:
```yaml
activated: yes
shims_on_path: yes
```
but when you `which python` you get the shim, so I'm guessing you activate first then activate shims second. e.g.
```shell
eval "$(mise activate bash)"
eval "$(mise activate bash --shims)"
```
I believe this is why you are seeing this issue.

If you looked at your path after your `mise trust` you'll likely see this order:
```console
$ echo $PATH | sed 's/ /\n/g'
/home/tecoholic/.local/share/mise/shims
/home/tecoholic/code/test/.venv/bin
/home/tecoholic/.local/share/mise/installs/python/3.11.9/bin
... (more omitted paths)
```
Which means when you run the `which python` it hits the shims directory before the .venv directory.
If you switch the order of activations it should fix your paths:
```shell
eval "$(mise activate bash --shims)"
eval "$(mise activate bash)"
```
You should get a better path order:
```console
$ echo $PATH | sed 's/ /\n/g'
/home/tecoholic/code/test/.venv/bin
/home/tecoholic/.local/share/mise/installs/python/3.11.10/bin
... (more /home/tecoholic/.local/share/mise/installs paths)
/Users/tecoholic/.local/share/mise/shims
... (more other paths)
```
And `which` should resolve to the .venv directory first
```console
$ which python
/home/tecoholic/code/test/.venv/bin/python
```
However `/home/tecoholic/code/test/.venv/bin/python` will always be a symlink to `/home/tecoholic/.local/share/mise/installs/python/3.11.10/bin/python` anyway, but this is expected behaviour. The .venv will  be the location all of your pip installed packages and bins.

Let me know if this fixes your issue

**Comment by tecoholic** [comment]
@syhol Thank you for the amazing, detailed and accurate explanation of the issue. I was surprised that I had both modes activated as my `.bashrc` had only the `activate` and no shims. Then I discovered the `--shims` in the `.profile` which I seemed to have added to make the IDE's happy. So, I had it the wrong way around. Interactive shells were adding shims and non-interactive shells were loading activate. I switched it and now everything is working as expected.

Again, thanks a lot for a great response.
